THIS FILE SHOULD BE RUN ON YOUR LOCAL(JUPYTER NOTEBOOK)

Extracting all text by using selenium library

In [2]:
#Importing packages
from selenium import webdriver
import pandas as pd

You need to download chromedriver then copy-paste chromedriver.exe path to the line below

In [3]:
driver = webdriver.Chrome(r"C:\Users\MONSTER\Desktop\driver\chromedriver.exe")

In [4]:
driver.get('https://www.nanos.ai/')

In [5]:
#/html/body---> Xpath

In [6]:
text = driver.find_elements_by_xpath("/html/body")[0]
all_text = text.text
all_text

'Search:\nNanos is a do-it-yourself\npay-as-you-go\nadvertising tool\nCreate optimized ad campaigns on Google, Facebook, and Instagram all in one place with Nanos patented Artificial Intelligence technology\nWatch How It Works\nBecome your own marketing agency in 4 easy steps\nNo website, no Facebook Business page required\nOnline marketing should be this simple\nIt’s like having an agency in your pocket!\nStep 1: Sign up at Nanos and tell us in a few words about your great product or service\nStep 2: Nanos AI will suggest ad text, keywords, interests, and platforms for your ad\nStep 3: Decide on the budget you want to spend, and how long to run your ads\nStep 4: Log in to your dashboard and watch Nanos AI in action optimizing at lightning speed\nAs seen on:\nThe digital marketing superhero for every kind of business\nWhether you have a physical store or an online venture, the goal is the same: to grow your business!\nWe understand that existing marketing tools may be too complicated a

import NLP libraries 

In [7]:
#from textblob import TextBlob
from sklearn import model_selection, preprocessing, linear_model, naive_bayes, metrics
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn import decomposition, ensemble
from nltk.tokenize import sent_tokenize, word_tokenize
import nltk
nltk.download("punkt")
from nltk.stem import WordNetLemmatizer
from nltk.corpus import stopwords
from warnings import filterwarnings
filterwarnings('ignore')
from gensim.models import Word2Vec

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\MONSTER\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


lenght of our all_text

In [8]:
len(all_text)

3707

tokenize word

In [9]:
text=word_tokenize(all_text.lower())

only alphabetical strings

In [10]:
text = [w for w in text if w.isalpha()]

In [11]:
stop_words = stopwords.words('english')

meaningful words that come from stopwords

In [12]:
text = [t for t in text if t not in stop_words]

meaningful text lenght

In [13]:
len(text)

359

Word Embedding

In [14]:
import numpy as np

In [15]:
from tensorflow.keras.models import Sequential,Model

In [16]:
from tensorflow.keras.layers import Embedding

In [17]:
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences

In [18]:
MAX_NB_WORDS = 100
MAX_SEQUENCE_LENGTH = 39

In [19]:
text[:2]

['search', 'nanos']

modelling

In [20]:
tokenizer = Tokenizer(nb_words=MAX_NB_WORDS)
tokenizer.fit_on_texts(text)
sequences = tokenizer.texts_to_sequences(text)

word_index = tokenizer.word_index
print('Found %s unique tokens.' % len(word_index))

Found 237 unique tokens.


In [21]:
word_index

{'nanos': 1,
 'marketing': 2,
 'business': 3,
 'online': 4,
 'ad': 5,
 'technology': 6,
 'campaign': 7,
 'ai': 8,
 'ads': 9,
 'agency': 10,
 'step': 11,
 'platforms': 12,
 'advertising': 13,
 'create': 14,
 'facebook': 15,
 'one': 16,
 'page': 17,
 'like': 18,
 'us': 19,
 'product': 20,
 'keywords': 21,
 'interests': 22,
 'budget': 23,
 'success': 24,
 'tool': 25,
 'google': 26,
 'instagram': 27,
 'watch': 28,
 'become': 29,
 'website': 30,
 'sign': 31,
 'great': 32,
 'text': 33,
 'run': 34,
 'optimizing': 35,
 'landing': 36,
 'time': 37,
 'clients': 38,
 'agencies': 39,
 'small': 40,
 'stories': 41,
 'make': 42,
 'thanks': 43,
 'use': 44,
 'effective': 45,
 'terms': 46,
 'blog': 47,
 'jan': 48,
 'read': 49,
 'article': 50,
 'contact': 51,
 'free': 52,
 'search': 53,
 'optimized': 54,
 'campaigns': 55,
 'place': 56,
 'patented': 57,
 'artificial': 58,
 'intelligence': 59,
 'works': 60,
 'easy': 61,
 'steps': 62,
 'required': 63,
 'simple': 64,
 'pocket': 65,
 'tell': 66,
 'words': 67,


 load the whole embedding into memory, I used glove trained model

In [22]:

embeddings_index = dict()
f = open('./glove.6B.100d.txt', encoding="utf-8")
for line in f:
	values = line.split()
	word = values[0]
	coefs = np.asarray(values[1:], dtype='float32')
	embeddings_index[word] = coefs
f.close()

len(embeddings_index)

Adding 1 because of reserved 0 index. Later we will reduce this zero.

In [23]:
vocab_size = len(tokenizer.word_index) + 1
print (vocab_size)

238


create a weight matrix for words in training docs

In [24]:
embedding_matrix = np.zeros((vocab_size, 100))
for word, i in tokenizer.word_index.items():
	embedding_vector = embeddings_index.get(word)
	if embedding_vector is not None:
		embedding_matrix[i] = embedding_vector

In [25]:
tokenizer.word_index.get("nanos")

1

In [26]:
tokenizer.word_index.get("marketing")

2

In [27]:
embedding_matrix.shape

(238, 100)

nanos embedding_matrix

In [28]:
embedding_matrix[1]

array([-0.50895   , -0.62186998, -0.17041001, -0.59547001,  0.14644   ,
       -0.56857997,  0.67795002,  0.18999   ,  0.53547001,  0.22712   ,
       -0.45181999,  1.05700004, -0.19568001, -0.025027  ,  0.12122   ,
       -0.37439001, -0.22944   ,  0.73356998,  0.75897002, -0.16356   ,
       -0.16004001, -0.18779001,  0.30711001, -0.03862   ,  0.12699001,
        0.52557999,  0.44400999,  0.24555001,  0.19909   ,  0.28669   ,
       -0.41064   ,  0.052381  , -0.34483001,  0.56068999,  0.05825   ,
        0.26339   , -0.21213   ,  0.15447   ,  0.10869   , -0.61321998,
        0.31114   ,  0.61161   , -0.48236999,  0.11865   ,  0.24201   ,
        0.62945998,  0.13932   , -0.40253001, -0.32706001,  0.92935002,
        0.05542   , -0.91016001, -0.24862   , -0.78986001, -0.65316999,
        0.42179   ,  0.20321999, -0.22351   ,  0.14002   , -0.53323001,
        0.47420001,  0.048665  ,  0.063908  ,  0.19746999, -0.60047001,
        0.57744998, -0.097913  , -0.045353  , -0.11291   , -0.33

In order to find similarity of words, using cosine_similarity

In [29]:
from sklearn.metrics.pairwise import cosine_similarity
import numpy as np

In [30]:
sim_mat = cosine_similarity(embedding_matrix,embedding_matrix)

In [31]:
sim_mat.shape

(238, 238)

similarity matrix for nanos

In [32]:
sim_mat[1]

array([ 0.        ,  1.        , -0.03635012, -0.2201808 , -0.0492457 ,
       -0.15194036,  0.01661878, -0.21902691,  0.12980407, -0.12631538,
       -0.14582046, -0.11327082,  0.07147027, -0.1608474 , -0.05334705,
        0.03649543, -0.1991473 , -0.12575312, -0.18846528, -0.19289147,
       -0.05321964, -0.12914448, -0.32873311, -0.08747623, -0.17385242,
       -0.1256644 ,  0.06687981,  0.14940334, -0.07708942, -0.13231209,
       -0.00832716, -0.16387112, -0.22171543, -0.14559109, -0.17089942,
        0.02058547, -0.04762698, -0.28719919, -0.2577703 , -0.21379487,
       -0.1072432 , -0.31826807, -0.11614724, -0.20130063, -0.10468696,
       -0.11668682, -0.23278099,  0.00761174, -0.07999595, -0.13757201,
       -0.2381364 , -0.31831065, -0.12971017, -0.15133994,  0.09747418,
       -0.23959899, -0.24857704,  0.12760671,  0.10023005, -0.1373164 ,
       -0.12258395, -0.16192208, -0.15769151, -0.16958424, -0.14655286,
       -0.04042769, -0.17465081, -0.1804527 , -0.1432407 , -0.14

In [33]:
#We should get rid of first item which is zero. The rest of them are similarity with "tv"
sim_mat[1][1:]

array([ 1.        , -0.03635012, -0.2201808 , -0.0492457 , -0.15194036,
        0.01661878, -0.21902691,  0.12980407, -0.12631538, -0.14582046,
       -0.11327082,  0.07147027, -0.1608474 , -0.05334705,  0.03649543,
       -0.1991473 , -0.12575312, -0.18846528, -0.19289147, -0.05321964,
       -0.12914448, -0.32873311, -0.08747623, -0.17385242, -0.1256644 ,
        0.06687981,  0.14940334, -0.07708942, -0.13231209, -0.00832716,
       -0.16387112, -0.22171543, -0.14559109, -0.17089942,  0.02058547,
       -0.04762698, -0.28719919, -0.2577703 , -0.21379487, -0.1072432 ,
       -0.31826807, -0.11614724, -0.20130063, -0.10468696, -0.11668682,
       -0.23278099,  0.00761174, -0.07999595, -0.13757201, -0.2381364 ,
       -0.31831065, -0.12971017, -0.15133994,  0.09747418, -0.23959899,
       -0.24857704,  0.12760671,  0.10023005, -0.1373164 , -0.12258395,
       -0.16192208, -0.15769151, -0.16958424, -0.14655286, -0.04042769,
       -0.17465081, -0.1804527 , -0.1432407 , -0.14117953, -0.13

In [34]:
for key, value in word_index.items() :
    print (key,value)

nanos 1
marketing 2
business 3
online 4
ad 5
technology 6
campaign 7
ai 8
ads 9
agency 10
step 11
platforms 12
advertising 13
create 14
facebook 15
one 16
page 17
like 18
us 19
product 20
keywords 21
interests 22
budget 23
success 24
tool 25
google 26
instagram 27
watch 28
become 29
website 30
sign 31
great 32
text 33
run 34
optimizing 35
landing 36
time 37
clients 38
agencies 39
small 40
stories 41
make 42
thanks 43
use 44
effective 45
terms 46
blog 47
jan 48
read 49
article 50
contact 51
free 52
search 53
optimized 54
campaigns 55
place 56
patented 57
artificial 58
intelligence 59
works 60
easy 61
steps 62
required 63
simple 64
pocket 65
tell 66
words 67
service 68
suggest 69
decide 70
want 71
spend 72
long 73
log 74
dashboard 75
action 76
lightning 77
speed 78
seen 79
digital 80
superhero 81
every 82
kind 83
whether 84
physical 85
store 86
venture 87
goal 88
grow 89
understand 90
existing 91
tools 92
may 93
complicated 94
expensive 95
everyone 96
expert 97
afford 98
hire 99
got 100


Create Data Frame in pandas

In [35]:
df = pd.DataFrame.from_dict(word_index, orient='index')
df=df.rename(columns={0: "similarity"})

Now similarity values come from index numbers

In [36]:
df.head()

similarity
nanos               1
marketing           2
business            3
online              4
ad                  5

In [37]:
word_index['digital']

80

index number of 'digital' is 80

In [38]:
df.similarity=sim_mat[80][1:]

In [39]:
df.sort_values('similarity', ascending = False)[:20]

similarity
digital        1.000000
video          0.739138
technology     0.660551
online         0.648975
network        0.614934
access         0.583072
videos         0.556028
media          0.554436
interface      0.541533
use            0.539343
advertising    0.538467
create         0.531965
product        0.525318
google         0.520686
tools          0.518354
marketing      0.504868
multiple       0.502196
text           0.501739
tool           0.498788
business       0.492566

As we see above, digital similarity is 0.660551 for technology, is 0.504868 for marketing

by calling word_sim function that is below, you can easily find any word similarity

In [41]:
def word_sim(word):
    index = tokenizer.word_index.get(word)
    df = pd.DataFrame.from_dict(word_index, orient='index')
    df=df.rename(columns={0: "similarity"})
    df.similarity=sim_mat[index][1:]
    df = df.sort_values('similarity', ascending = False)[:10]
    return df

pick a word

In [42]:
word_index.keys()

dict_keys(['nanos', 'marketing', 'business', 'online', 'ad', 'technology', 'campaign', 'ai', 'ads', 'agency', 'step', 'platforms', 'advertising', 'create', 'facebook', 'one', 'page', 'like', 'us', 'product', 'keywords', 'interests', 'budget', 'success', 'tool', 'google', 'instagram', 'watch', 'become', 'website', 'sign', 'great', 'text', 'run', 'optimizing', 'landing', 'time', 'clients', 'agencies', 'small', 'stories', 'make', 'thanks', 'use', 'effective', 'terms', 'blog', 'jan', 'read', 'article', 'contact', 'free', 'search', 'optimized', 'campaigns', 'place', 'patented', 'artificial', 'intelligence', 'works', 'easy', 'steps', 'required', 'simple', 'pocket', 'tell', 'words', 'service', 'suggest', 'decide', 'want', 'spend', 'long', 'log', 'dashboard', 'action', 'lightning', 'speed', 'seen', 'digital', 'superhero', 'every', 'kind', 'whether', 'physical', 'store', 'venture', 'goal', 'grow', 'understand', 'existing', 'tools', 'may', 'complicated', 'expensive', 'everyone', 'expert', 'affor

run word_sim function within 'marketing'

In [44]:
word_sim('marketing')

similarity
marketing      1.000000
advertising    0.818577
business       0.776054
product        0.698289
brand          0.686470
company        0.686448
strategy       0.679744
sales          0.663065
venture        0.631109
technology     0.606919

As we see above, the biggest similarity for marketing is advertising

In [45]:
word_sim('technology')

similarity
technology    1.000000
business      0.695139
digital       0.660551
tools         0.648552
tool          0.635280
design        0.616072
marketing     0.606919
based         0.599098
venture       0.595859
use           0.595067

As we see above, the highest similarity for technology is 0.695139